### ISSUE TITLE
ENOSPC: System limit for number of file watchers reached

### ISSUE THREAD

**Issue by quintenkasteel** [lydell/elm-watch#106]
We have a pretty large repo of elm files. If i make changes to my files i get the following error: 

✅ 11:25:12 Compilation finished in 517 ms.
-- WATCHER ERROR --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

The file watcher encountered an error, which means that it cannot continue.
elm-watch is powered by its file watcher, so I have to exit at this point.

See if this is something you can solve by maybe removing some problematic files
or something!

This is the error message I got:
ENOSPC: System limit for number of file watchers reached, watch ../.stack-work/install/../*.hs

the error occurs on every change, it looks like its adding file watchers for files not in the tree of elm and non elm files.

**Comment by lydell** [comment]
Hi!

Can you try `npm install elm-watch@beta` and see if it happens there too?

Just wanted to check if this is the same issue as https://github.com/lydell/elm-watch/issues/62. (That has been fixed on the main branch too since then, but I haven’t cut a new release yet.)

**Comment by quintenkasteel** [comment]
Hi, 

Thanks for your response! Unfortunatly it doesn't seem to fix the issue. It seems the fetch files that are not in the "directory tree" of the input files in elm-watch.json. The folder that it's mentioning is in a other directory then the input files. I can't provide any example of the code as it's a production app we are using for our company.

**Comment by lydell** [comment]
I see, thanks for trying that out!

elm-watch watches the following directories recursively (note: this is how it works on the beta version and the unreleased main branch):

1. The directory that contains elm-watch.json.
2. elm-watch finds the closes elm.json for every target and watches the directory of each elm.json.
3. elm-watch reads "source-directories" from each of those elm.json files and watches all of those too.

Is there any of those that are “higher up” and causes elm-watch to venture into `../.stack-work/`?

If you set the environment variables `__ELM_WATCH_DEBUG` and  `__ELM_WATCH_NOT_TTY` to anything, you’ll see a print when elm-watch starts that shows the whole “project” as elm-watch sees it, with all paths, including the watch root(s) (one or many depends on if you use the beta or not).

**Comment by quintenkasteel** [comment]
Here some additional info: 
```
GotWatcherEvent
{
  tag: 'GotWatcherEvent',
  date: 2025-01-07T13:31:28.172Z,
  eventName: 'changed',
  absolutePathString: '../.git/FETCH_HEAD'
}
{
  nextAction: { tag: 'NoAction' },
  hotState: { tag: 'Idle' },
  latestEvents: []
}
[ { tag: 'NoCmd' } ]
GotWatcherEvent
{
  tag: 'GotWatcherEvent',
  date: 2025-01-07T13:31:28.331Z,
  eventName: 'changed',
  absolutePathString: '../.stack-work/dist/x86_64-linux-tinfo6/ghc-8.8.4/stack-setup-config-mod'
}
{
  nextAction: { tag: 'NoAction' },
  hotState: { tag: 'Idle' },
  latestEvents: []
}
[ { tag: 'NoCmd' } ]
-- WATCHER ERROR --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

The file watcher encountered an error, which means that it cannot continue.
elm-watch is powered by its file watcher, so I have to exit at this point.

See if this is something you can solve by maybe removing some problematic files
or something!

This is the error message I got:

ENOSPC: System limit for number of file watchers reached
```

**Comment by lydell** [comment]
Thanks! It might help if I configure the watcher to only watch `*.elm`, `elm.json`, `elm-watch.json`, and removal of `elm-stuff/` then … oh – and the `postprocess.js` if you have one, that might be tricky 🤔  Maybe it makes more sense to add a hardcoded exclusion for `.stack-work` just like I have hardcoded ignoring for `node_modules`. I’m planning to rewrite the watcher anyway in a future version, so that might be good enough to unblock you for now.

**Comment by lydell** [comment]
Could you help me and test one thing?

1. Edit `node_modules/elm-watch/index.js`.
2. Search for `node_modules`.
3. Make the below change:

```diff
-    ignored: /\/(elm-stuff|node_modules)\//,
+    ignored: /\/(elm-stuff|node_modules|\.stack-work)\//,
```

Does it make a difference? If it does, I could get that quickfix out for the next release. A more proper solution would take longer.

**Comment by quintenkasteel** [comment]
Works like a charm!

**Comment by lydell** [comment]
v1.2.0 has been released. It ignores `.stack-work/`. Keeping the issue open since I intend to solve this in a better way in the future.

**Comment by anka-213** [comment]
Can we make it configurable or do you have an even better plan?

**Comment by lydell** [comment]
@anka-213 I have another plan. But I haven’t been good at all finishing my plans lately.

What do you need to exclude?

**Comment by anka-213** [comment]
I have a project specific folder called `_cache` with hundreds of thousands of json files.

**Comment by thirafidide** [comment]
@lydell Just want to add that I keep getting OOM after a while because turns out it follows .devenv symlinks. For now I just patching the ignore list to add .devenv

**Comment by lydell** [comment]
Thanks! Would you mind checking if `elm-test --watch` (version 0.19.2-1) has the same problem?

**Comment by thirafidide** [comment]
i don't see any problem with `elm-test --watch`, since it reads src/ instead project root (where .devenv lives) because `"source-directories": ["src"]`, but looks like it will still be a problem if there is symlinks inside src/ for some reasons
